README.md
# divergingPips

### Description

R package for creation of "diverging pip" plots

I designed these figures to help identify imbalances between two groups where both percentages and raw frequencies were important. For instance, if one sees a 70/30 split between two groups with 300 people, this might be a problem; but if the number of people is 10, that is less of a problem. Typical plots will show percentages with numerical frequencies added; these plots are confusing and don't represent the data very well. 

There are four kinds of plots currently supported: plain, side-by-side, stacked, and histogram-style.
![Example figure](http://richarddmorey.org/images/div_pips_all.png)

| Type | Example use          |
| ------------- | ----------- |
| plain      | Show group balance across time. |
| side-by-side     | Compare group balance across time between two or more populations |
| stacked     | Compare group balance in, say, turnover across time |
| hisstogram-style     | Compare group balance in distribution of ordinal variable (e.g., grades) |

### How to use

To install, use the `devtools` package:
    
    install.packages('devtools') # if necessary
     
    devtools::install_github("richarddmorey/divergingPips", subdir = "divergingPips")

Then run the examples:
    
    example("diverging_pip_plot", "divergingPips")

Or try this example:

    frq = array( c(10,40,20,10,
                   30,20,15,20,
                   15,15,20,22,
                   11,23,54,23), dim = c(2,4,2))
    
    rownames(frq) = c("Women","Men")
    colnames(frq) = 1981:1984
                       
    cols = matrix(c(rgb(255,164,56,maxColorValue = 255),
                rgb(120,80,155,maxColorValue = 255),
                rgb(255,164,56,maxColorValue = 255, alpha=128),
                rgb(120,80,155,maxColorValue = 255, alpha=128)),2,2)
    
    divergingPips::diverging_pip_plot(frq, bar.width = .4, bar.width.n = 5, bar.col = cols, sym = FALSE, stacked = TRUE)  

which results in this figure:
![Example figure](http://richarddmorey.org/images/div_pips.png)


NEWS
CHANGES IN divergingPips 0.1:

  * Initial version. Single function: diverging_pip_plot (see examples)


divergingPips/R/add_pip_bar.R
#' Add a series of pip bars to a plot 
#'
#' @param x array of x axis values where to center the bars
#' @param ns array of frequencies
#' @param bar.width The width of a bar in x units
#' @param bar.width.n The number of columns of pips to break the bars into
#' @param line.col A vector of length 2 giving the colors of the pip lines for the top and bottom bars
#' @param hist Create histogram-style bars?
#' @param bar.col A vector or array giving the colors of the bars (see details)
#' @param line.wd Width of the pip lines
#' @param top.cex The magnification of the top labels relative to par(cex.axis)
#' @param ... Unused additional arguments
#'
#' @return Returns NULL invisibly.
add.pip.bar <- function(x, ns, 
                         bar.width = .5,
                         bar.width.n = 2, 
                         line.col = c("white", "white"), 
                         bar.col = c("red", "blue"),
                         hist = FALSE, line.wd = 1, 
                         top.cex = top.cex, ...)
{
  
  if(length(dim(ns))!=2){
    dim(ns) = c(2,length(ns)/2)
  }
  
  if(hist & dim(ns)[2]>1){
    return(add.pip.bar.hist(x, ns, 
                             bar.width,
                             bar.width.n, 
                             line.col, bar.col,
                             line.wd = line.wd,
                             top.cex = top.cex, ...))
  } 
  
  
  if(length(dim(ns))>2){
    bar.col = matrix(bar.col,2,dim(ns)[3])
  }else{
    bar.col = matrix(bar.col,nrow=2)
  }
  
  
  bar.height.n.all = ceiling(rowSums(ns) / bar.width.n)
  bar.leftover.n = rowSums(ns) %% bar.width.n
  bar.height.n = rowSums(ns) %/% bar.width.n
  
  for(i in 1:2){
    bar.matrix = matrix(NA, bar.width.n, bar.height.n.all[i])
    for(j in 1:dim(ns)[2]){
      if(ns[i,j]==0){
        next
      }
      idx = c(0,cumsum(ns[i,]))[j:(j+1)] + c(1,0)
      bar.matrix[idx[1]:idx[2]]=j
      y.b = apply(bar.matrix==j,1,function(v) range(which(v))) - c(1,0)
      y.b[is.infinite(y.b)]=0
      y.b = c(y.b[1,], rev(y.b[2,]))
      y.c = 2*(2 - i - .5)*rep(y.b,each=2)
      x.c = seq(x - bar.width/2, x + bar.width/2, bar.width/bar.width.n)
      x.c = c(x.c[1],  rep(x.c[-c(1,bar.width.n+1)],each=2) ,x.c[bar.width.n+1])
      x.c = c(x.c,rev(x.c))
      polygon(x.c,y.c,border=line.col[i],col=bar.col[i,j], lwd = line.wd)
    }
  }
  
  for(i in 1:2){
    ## Draw segments
    for(j in 1:(bar.width.n-1)){
      segments(j*bar.width/bar.width.n + x - bar.width/2,
               0,
               j*bar.width/bar.width.n + x - bar.width/2,
               2*(2 - i - .5)*bar.height.n[i],
               col=line.col[i], lwd = line.wd)
    }
    
    if( bar.leftover.n[i] > 1){
      for(j in 1:(bar.leftover.n[i]-1)){
        segments(j*bar.width/bar.width.n + x - bar.width/2,
                 2*(2 - i - .5)*bar.height.n.all[i],
                 j*bar.width/bar.width.n + x - bar.width/2,
                 2*(2 - i - .5)*bar.height.n[i],
                 col=line.col[i], lwd = line.wd)
      }
    }
    if(bar.height.n.all[i]>1){
      for(j in 1:bar.height.n[i]){
        segments(x - bar.width/2,
                 2*(2 - i - .5)*j,
                 x + bar.width/2,
                 2*(2 - i - .5)*j,
                 col=line.col[i], lwd = line.wd)
      }
    }
  }  
  invisible(NULL)
}





#' Add a histogram-style series of pip bars to a plot 
#'
#' @param x array of x axis values where to center the bars
#' @param ns array of frequencies
#' @param bar.width The width of a bar in x units
#' @param bar.width.n The number of columns of pips to break the bars into
#' @param line.col A vector of length 2 giving the colors of the pip lines for the top and bottom bars
#' @param line.wd Width of the pip lines
#' @param bar.col A vector or array giving the colors of the bars (see details)
#' @param top.cex The magnification of the top labels relative to par(cex.axis)
#' @param ... Unused additional arguments
#'
#' @return Returns NULL invisibly.
add.pip.bar.hist <- function(x, ns, 
                              bar.width = .5,
                              bar.width.n = 2, 
                              line.col = c("white", "white"), 
                              bar.col = c("red", "blue"),
                              line.wd = 1, 
                              top.cex = 1, ...)
{
  
  if(length(dim(ns))>2){
    bar.col = matrix(bar.col,2,dim(ns)[3])
  }else{
    bar.col = matrix(bar.col,nrow=2)
  }
  
  
  bar.height.n.all = ceiling(ns / bar.width.n)
  bar.leftover.n = ns %% bar.width.n
  bar.height.n = ns %/% bar.width.n
  total.width = bar.width * dim(ns)[2]
  
  for(j in 2:dim(ns)[2]){
    abline(v=x - total.width/2 + (j-1)*bar.width,col="lightgray",lty=3) 
  }
  
  for(i in 1:2){
    for(j in 1:dim(ns)[2]){
      x0 = x - total.width/2 + (j-1)*bar.width + bar.width/2
      text(x0, par()$usr[4],
           dimnames(ns)[[2]][j],cex=par()$cex.axis*top.cex,adj=c(.5,1.2))
      bar.matrix = matrix(NA, bar.width.n, bar.height.n.all[i,j])
      if(ns[i,j]==0){
        next
      }
      bar.matrix[1:ns[i,j]] = 1
      if(j<=dim(ns)[2]/2){
        bar.matrix = bar.matrix[bar.width.n:1,]
      }
      bar.matrix = matrix(bar.matrix,nrow=bar.width.n)
      y.b = apply(bar.matrix==1,1,function(v) range(which(v))) - c(1,0)
      y.b[is.infinite(y.b)]=0
      y.b = c(y.b[1,], rev(y.b[2,]))
      y.c = 2*(2 - i - .5)*rep(y.b,each=2)
      x.c = seq(x0 - bar.width/2, x0 + bar.width/2, bar.width/bar.width.n)
      x.c = c(x.c[1],  rep(x.c[-c(1,bar.width.n+1)],each=2) ,x.c[bar.width.n+1])
      x.c = c(x.c,rev(x.c))
      polygon(x.c,y.c,border=line.col[i],col=bar.col[i,j], lwd = line.wd)
      
      # Draw segments
      for(k in 1:(bar.width.n-1)){
        segments(k*bar.width/bar.width.n + x0 - bar.width/2,
                 0,
                 k*bar.width/bar.width.n + x0 - bar.width/2,
                 2*(2 - i - .5)*bar.height.n[i,j],
                 col=line.col[i], lwd = line.wd)
      }
      if( bar.leftover.n[i,j] > 1){
        for(k in 1:(bar.leftover.n[i,j]-1)){
          s = 1-2*(j<=dim(ns)[2]/2)
          segments(s*k*bar.width/bar.width.n + x0 - s*bar.width/2,
                   2*(2 - i - .5)*bar.height.n.all[i,j],
                   s*k*bar.width/bar.width.n + x0 - s*bar.width/2,
                   2*(2 - i - .5)*bar.height.n[i,j],
                   col=line.col[i], lwd = line.wd)
        }
      }
      if(bar.height.n[i,j]>0){
        for(k in 1:(bar.height.n[i,j])){
          segments(x0 - bar.width/2,
                   2*(2 - i - .5)*k,
                   x0 + bar.width/2,
                   2*(2 - i - .5)*k,
                   col=line.col[i], lwd = line.wd)
        }
      }
      
    }
  }
  invisible(NULL)
}




divergingPips/R/diverging_pip_plot.R

#' Create a diverging pip bar plot from frequencies in an array
#' 
#' 
#'
#' @param ns A 2 by N by M array (see details)
#' @param bar.width The width of a bar in x units
#' @param bar.width.n The number of columns of pips to break the bars into
#' @param line.col A vector of length 2 giving the colors of the pip lines for the top and bottom bars
#' @param line.wd Width of the pip lines
#' @param bar.col A vector or array giving the colors of the bars (see details)
#' @param xlab x axis label
#' @param box Create box around entire plot?
#' @param tick.every Make a y axis tick after this many pip rows
#' @param sym Should the y axis by symmetric around 0?
#' @param cluster.width For side-by-side plots, width of all bars together in x axis units
#' @param panel.lty The line type for dividing lines across bars/clusters
#' @param stacked If TRUE, create a stacked bar plot (see details)
#' @param hist If TRUE, create a histogram-style bar chart (see details)
#' @param top.cex The magnification of the top labels relative to par(cex.axis)
#' 
#' @return Returns NULL invisibly.
#' @export
#'
#' @examples
#' ## A basic bar plot
#' frq = array( c(10,40,20,10,30,20), dim = c(2,4))
#' 
#' rownames(frq) = c("Women","Men")
#' colnames(frq) = 1995:1998
#' 
#' diverging_pip_plot(frq, bar.width = .3, bar.width.n = 4, sym = TRUE)
#' 
#' ## Side-by-side bar plot
#' frq = array( c(10,40,20,10,
#'                30,20,15,20,
#'                15,15,20,22,
#'                11,23,54,23), dim = c(2,4,2))
#' 
#' rownames(frq) = c("Women","Men")
#' colnames(frq) = 1995:1998
#' dimnames(frq)[[3]] = c("a","b")
#' 
#' bar.col = array(c("red","blue","darkred","darkblue"),dim=c(2,2))
#' 
#' diverging_pip_plot(frq, bar.width = .2, bar.width.n = 4, bar.col = bar.col,
#'                    sym = TRUE, cluster.width = .4, panel.lty = 1)
#' 
#' ## Stacked bar plot
#' diverging_pip_plot(frq, bar.width = .5, bar.width.n = 6, bar.col = bar.col,
#'                    sym = TRUE, cluster.width = .2, panel.lty = 1, stacked=TRUE)
#' 
#' ## Histogram-style plot
#' 
#' bar.col = array(c("pink","lightblue","red","blue","darkred","darkblue"),dim=c(2,3))
#' 
#' frq = array( c(10,40,20,10,
#'                30,20,15,20,
#'                15,15,20,22,
#'                11,23,54,23,
#'                20,15,20,15,
#'                15,20,22,31), dim = c(2,4,3))
#' 
#' rownames(frq) = c("Women","Men")
#' colnames(frq) = 1995:1998
#' dimnames(frq)[[3]] = c("a","b","c")
#' 
#' diverging_pip_plot(frq, bar.width = .2, bar.width.n = 6, bar.col = bar.col,
#'                    sym = TRUE, cluster.width = .3, panel.lty = 1, hist=TRUE)
diverging_pip_plot <- function(ns, bar.width = .3,
                            bar.width.n = 2, 
                            line.col = c("white", "white"), 
                            line.wd = 1,
                            bar.col = c("red", "blue"),
                            xlab="", box=TRUE, tick.every=5, sym=FALSE,
                            cluster.width=.3, panel.lty=0,
                            stacked=FALSE, hist=FALSE,top.cex = 1)
{
  
  if(length(dim(ns))>2){
    bar.col = matrix(bar.col,2,dim(ns)[3])
  }
  
  if(length(ns)==2) ns = matrix(ns,2,1)
  if(length(dim(ns))==3){
    side.by.side=TRUE
    n.in.cluster = dim(ns)[3]
    if(stacked & hist) stop("Only one of stacked and hist may be true at once.")
  }else{
    side.by.side=FALSE
    stacked = FALSE
    hist = FALSE
  }
  
  if(sym){
    if(stacked){
      ymax = max(ceiling(apply(ns,c(1,2),sum)/bar.width.n))
    }else{
      ymax = max(ceiling(ns/bar.width.n))
    }
  }else{
    if(stacked){
      ymax = rev(apply( ceiling(apply(ns,c(1,2),sum)/bar.width.n), 1, max ))
    }else{
      ymax = rev(apply( ceiling(ns/bar.width.n), 1, max ))
    }
  }
  ylims = c(-1,1)*ymax*1.05
  
  if(side.by.side){
    xlims = c(.5, ncol(ns) + .5)
  }else{
    xlims = c(.8, ncol(ns) + .2)
  }
  
  
  par(las=1)
  plot(0,0,xlim=xlims,ylim=ylims,typ='n',axes=FALSE,ylab="",xlab=xlab)
  
  for(i in 1:ncol(ns)){
    if(side.by.side){
      if(stacked | hist){
        add.pip.bar(i,ns[,i,],bar.width,
                bar.width.n, 
                line.col, 
                bar.col,
                hist = hist, line.wd = line.wd, 
                top.cex = top.cex)
      }else{
        for(j in 0:(n.in.cluster-1)){
          add.pip.bar(i - cluster.width/2 + j*cluster.width/(n.in.cluster-1),
                  ns[,i,j+1],bar.width,
                  bar.width.n, 
                  line.col, 
                  bar.col[,j+1],
                  line.wd = line.wd, top.cex = top.cex)
          text(i - cluster.width/2 + j*cluster.width/(n.in.cluster-1),
               par()$usr[4],
               dimnames(ns)[[3]][j+1],cex=par()$cex.axis*top.cex,adj=c(.5,1.2))
        }
      }
    }else{
      add.pip.bar(i,ns[,i],bar.width,
              bar.width.n, 
              line.col, 
              bar.col,
              line.wd = line.wd, top.cex = top.cex)
    }
  } 
  axis(1,at=1:ncol(ns),labels=colnames(ns),tick = FALSE)
  ticks = seq(tick.every, ceiling(max(ymax)/tick.every)*tick.every,tick.every)
  axis(2,at=ticks,labels=ticks*bar.width.n)
  axis(2,at=-ticks,labels=ticks*bar.width.n)
  
  for(i in 1:2){
    mtext(rownames(ns)[i],2,par()$mgp[1],cex=par()$cex.lab,adj=2-i,las=0)
  }
  if(dim(ns)[2]>1 & side.by.side){
    abline(v=seq(1.5,ncol(ns)-.5,1),col="gray",lty=panel.lty)
  }
  abline(h=0,col="gray",lty=1)
  
  if(box) box()
  
  invisible(NULL)
}
